src/banded/BandedMatrix.jl

```diff
@@ -127,7 +128,7 @@ brand
 function BandedMatrix{T}(A::AbstractMatrix, bnds::NTuple{2,Int}) where T
     (n,m) = size(A)
     (l,u) = bnds
-    ret = BandedMatrix{T}(undef, n, m, bnds...)
+    ret = _BandedMatrix(similar(A, T, (max(0,l+u+1),m)), n, l, u)
```

Comment:
This would be cleaner as:
```julia
BandedMatrix{T,typeof(A)}(undef, n, m, bnds...)
```

Reply:
Hmm, after thinking about it there is no reason that the type of the data matrix should match `typeof(A)`. And it feels like one would expect `BandedMatrix{Float64}(speye(10), (0,0)) isa `BandedMatrix{Float64,Matrix{Float64}}`.

So I think the best thing to do here is always return a `BandedMatrix{T,Matrix{T}}`, that is, go back to the original version:
```julia
ret = BandedMatrix{T}(undef, n, m, bnds...)
```

Reply:
But what about `BandedMatrix{Float64}(GPUArray(...),...)`? I would expect it resides on the GPU.

I guess since BandedMatrix is sparse, there could be a specialization for matrices derived from `AbstractSparseArray`.

Reply:
thats a tricky one: it probably won’t be efficient since it needs to call getindex. 

Note in general these constructors shouldn’t be used in high performance situations: the point of a banded matrix is to minimise data usage, so first creating a dense GPUArray to then create a sparse BandedMatrix is a bad idea

I’d say leave it to always create a ` BandedMatrix{T,Matrix{T}} ` and deal with this if it ever comes up as a performance issue.

Reply:
allright.